target_snowflake/sinks.py

```diff
@@ -1,44 +1,140 @@
 """Snowflake target sink class, which handles writing streams."""
 
-from typing import Dict, List, Optional
+import datetime
+import tempfile
+import pytz
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional
+from typing import Any, Dict, List
 
-from singer_sdk.target_base import Target
-from target_snowflake.database_target.csv_sink import CSVSink
-from target_snowflake.migrator import SnowflakeSchemaMigrator
+from singer_sdk import SQLSink, SQLTarget
 
+from target_snowflake.connector import SnowflakeConnector
+from target_snowflake.stages import NamedStage
 
-class SnowflakeSink(CSVSink):
+import csv
+
+
+class SnowflakeSink(SQLSink):
     """Snowflake target sink class."""
 
+    connector_class = SnowflakeConnector
+    stage_class = NamedStage
+
     def __init__(
         self,
-        target: Target,
+        target: SQLTarget,
         stream_name: str,
         schema: Dict,
         key_properties: Optional[List[str]],
     ) -> None:
         super().__init__(target, stream_name, schema, key_properties)
+        self.prepare_sink()
+
+    @property
+    def table_name(self) -> str:
+        return self.stream_name.split("-")[-1].upper()
+
+    @property
+    def schema_name(self) -> Optional[str]:
+        """Return the target schema name."""
+        return self.config["snowflake"]["schema"].upper()
+
+    @property
+    def database_name(self) -> Optional[str]:
+        """Return the target DB name."""
+        return self.config["snowflake"]["database"].upper()
 
-        self.connection = target.connect()
-        self._has_migrated = False
-        self.migrator = SnowflakeSchemaMigrator(sink=self)
-        self.table_schema = target.table_schema
+    def prepare_sink(self) -> None:
+        self.connector.connection.execute(
+            'CREATE SCHEMA IF NOT EXISTS "{self.schema_name}"'
+        )
+        self.connector.prepare_table(
+            self.full_table_name, schema=self.schema, primary_keys=self.key_properties
+        )
+        self.stage = self.stage_class(self)
 
     @property
     def max_size(self):
         return self.config["batch_size_rows"]
```

Comment:
This is probably fine for now, but Snowflake suggests for optimal loading you break your files into 100-250MB compressed. Allowing this would require some significant changes to `BatchSink` I think.

Reply:
@rabidaudio - This shouldn't (in theory) be too difficult. Instead of overriding max_size (which is a row count), we can override `is_full()`.

The default is_full implementation is:

>     @property
>     def is_full(self) -> bool:
>         """Check against size limit.
> 
>         Returns:
>             True if the sink needs to be drained.
>         """
>         return self.current_size >= self.max_size

So, if there were a way to estimate the file size (as opposed to row counts), you could put that logic here.